### ISSUE TITLE
Help required on Latest patchset applied to DBCS

### ISSUE THREAD

**Issue by dralquinta** [oracle/oci-python-sdk#412]
Hi Folks, 

I'm working on an implementation where I have to check if the latest and greatest available patchset is installed on DBCS DB Systems, however I'm not getting any straight forward way of doing it using the API's available. 

I'm currently using client: https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/database/client/oci.database.DatabaseClient.html

And I'm using the following functions: 

1. https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/database/client/oci.database.DatabaseClient.html#oci.database.DatabaseClient.list_db_home_patches
2. https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/database/client/oci.database.DatabaseClient.html#oci.database.DatabaseClient.list_db_systems
3. https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/database/client/oci.database.DatabaseClient.html#oci.database.DatabaseClient.list_db_homes 

My initial idea (and probably the right move) was to only get the patches and see if the JSON returned contained any alteration on entry `available_actions`. as depicted in point 1. My assumption here was that if the patch was already applied, then no further actions should be available up until a new patch is ready to go, however the JSON returned will always get this (regardless if any patch was applied or not, it'll always be either `APPLY` or `PRECHECK`): 

```json
{
  "available_actions": [
    "APPLY",
    "PRECHECK"
  ],
  "description": "Oct 2021 19c Database patch",
  "id": "ocid1.dbpatch.oc1.iad.anuwcljst5t4sqqawfxptoxhatfcu7loxrvkp3yo3t6s7ureljm6a2z6i2aq",
  "last_action": null,
  "lifecycle_details": null,
  "lifecycle_state": null,
  "time_released": "2021-10-19T01:00:00+00:00",
  "version": "19.13.0.0.211019"
}
```

Then I moved into using the DB_System call from point 2. This returns something somewhat more useful, however nothing that can really tell me exactly if it's the latest patch applied already. 

For a database that has any patch applied, then the value of `last_patch_history_entry_id` will return something as follows (cutted down json for practical purposes)

```json
{
  "availability_domain": "oDQF:SA-SANTIAGO-1-AD-1",
  "backup_network_nsg_ids": null,
  "backup_subnet_id": null,
  "cluster_name": null,
  "compartment_id": "ocid1.compartment.oc1..aaaaaaaalx77hsir4r2kc36twf3f3twf3ff3b5jtam7wbe5jmtq3bjzvjgzq",
  "cpu_core_count": 4,
  "data_storage_percentage": 80,
  "data_storage_size_in_gbs": 256,
  "database_edition": "ENTERPRISE_EDITION_HIGH_PERFORMANCE",
  "db_system_options": {
    "storage_management": "ASM"
  },
  "defined_tags": {
    "default_namespace": {
      "creator": "oracleidentitycloudservice/denny.alquinta@oracle.com"
    }
  },
  "disk_redundancy": "HIGH",
  "display_name": "DBSystem 202201041102",
  "domain": "dalquinthubpvts.dalquinthubvcn.oraclevcn.com",
  "fault_domains": [
    "FAULT-DOMAIN-2"
  ],
  "freeform_tags": {},
  "hostname": "foobar",
  "id": "ocid1.dbsystem.oc1.sa-santiago-1.anzwgljrtwfhi7yak552t2ca4plannrqnw47xlzmkvblvuuuqpip2hnmz2pa",
  "kms_key_id": null,
  "last_maintenance_run_id": null,
  "last_patch_history_entry_id": "ocid1.dbpatchhistory.oc1.sa-santiago-1.anzwgljrtwfhi7yaym6vx24ahsindpaqnqtopssm6ahhqip7btufsfrzp73a",
  "license_model": "LICENSE_INCLUDED",
  "lifecycle_details": null,
  "lifecycle_state": "AVAILABLE",
  "listener_port": 1521,
  "maintenance_window": null,
  "next_maintenance_run_id": null,
```

And for a DB that has no patches applied, then the value of `last_patch_history_entry_id` will be null as follows: 

```json
  "freeform_tags": {},
  "hostname": "dbsystem",
  "id": "ocid1.dbsystem.oc1.iad.anuwcljstwfhi7yajlim5dxmvjcbn36k6pxrdug6ghz65yanuwv2vm3ni7ba",
  "kms_key_id": null,
  "last_maintenance_run_id": null,
  "last_patch_history_entry_id": null,
```

This doesn't really help, as if I at least applied a patch once to all my DBs then this value will always contain something. 

Then I went for DB_HOME object from point 3. however here I really have no data that results useful at all for this purpose. 

My next move is to go with some crazy combination between patch history and other functions, however this still looks like a huge overkill for something that should be so simple, besides of the performance impact that this will have as I'll have to spawn multiple threads to consult all the regions in my tenancy and it'll turn to be really slow. 

I'm not really sure how to classify this issue, however first and foremost I'll say that this is somewhat a help request. Can you shed some light in what should I do? 

tldr; how to determine if a dbcs is in the latest and greatest available patchset by using Python SDK without building an overkill :) 

Thanks!

Denny.

**Comment by dralquinta** [comment]
Any chance on providing an update on this one? Thanks. 

Denny

**Comment by adizohar** [comment]
Hi Denny
The API only provides patch that applied using DBCLI or the console
If database patched using opatch or installed from patched db home, it won't be available in the API
For such, you will need to run opatch or query the database for latest patch
For your info

**Comment by dralquinta** [comment]
Hi Adi. 

In this case the patches are applied by using the console. Above explanation is based on this use case in particular.

Does that means that there is something weird with the API or maybe it's refresh rate?

Thanks

Denny

**Comment by dralquinta** [comment]
Another thing. 

In my case I really don't care about what patch is applied as long as it's the latest available. So in such sense if there is any shortcut that you guys can think of that I can take advantage from it'll be great.

**Comment by adizohar** [comment]
Hi Denny

You should use the get_db_home_patch_history_entry API
And get_db_home_patch to get the patch info

Below info from the API

"patches": [
	"Oct 2021 21c Database patch - 21.4.0.0.0 - 2021-10-19 - Last Action: None"
],
"patches_history": [
	"Jul 2021 21c Database patch 21.3.0.0.0 - SUCCEEDED - 2021-10-05 - 2021-10-05 - Last Action: APPLY",
	"Jul 2021 21c Database patch 21.3.0.0.0 - SUCCEEDED - 2021-10-02 - 2021-10-02 - Last Action: PRECHECK"
]

**Comment by dralquinta** [comment]
humm.. one thing there. Is there a way to know if the DB is on latest patchset anyways? I'm wondering this because I can totally get the patch history as I did above, however in order to know if it is the latest and greatest available I'd definitely need to do some sort of call to another place to check on the patch list (which is something I'm intentionally trying to avoid -  Something crazy like doing webscrapping of this site and hope for it to be always updated: https://docs.oracle.com/en-us/iaas/Content/Database/Tasks/patchingDB.htm ) 

So my hope was that this was covered around the available actions but I failed miserably with that assumption :(

**Comment by adizohar** [comment]
Hi Denny, the only way to check if database is on latest patch if no new patches exist in the list_db_home_patches API
